### PR TITLE
Chapter 13 - User microposts

### PR DESCRIPTION
**Pull request by EvonneNgXinYi**
Construct user microposts model for logged in user

### CODE REVIEW

**Comment by EvonneNgXinYi** [comment]
ready

File: app/controllers/microposts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  before_action :correct_user, only: :destroy
+
+  def create
+    @micropost = current_user.microposts.build micropost_params
+    @micropost.image.attach micropost_params[:image]
+    if @micropost.save
+      flash[:success] = t ".success_mess_created"
+    redirect_to root_path
+    else
+      @feed_items = current_user.feed.page(params[:page])
```
Comment (manhnt-0960) [inline app/controllers/microposts_controller.rb]: add number of per page

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -8,7 +8,10 @@ def index
     @users = User.sort_by_name.page(params[:page]).per(Settings.collections.users_default_page)
   end
 
-  def show; end
+  def show
+    @user = User.find params[:id]
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: please use method find_by instead find and don't forget handle case return nil :D

File: app/models/micropost.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+  validates :user_id, presence: true
+  validates :content, presence: true, length: { maximum: Settings.validations.content.max_length }
+  validates :image, content_type: { in: Settings.image.type, message: I18n.t("errors.messages.invalid_image_format") }, size: { less_than: Settings.image.size.megabytes, message: I18n.t("errors.messages.invalid_image_size") }
+  scope :recent_posts, -> { order created_at: :desc }
```
Comment (manhnt-0960) [inline app/models/micropost.rb:8]: add 1 empty line before this line

File: config/locales/en.yml [hunk 4]
```diff
@@ -147,3 +152,32 @@ en:
...
+      fail_mess_created: "Fail to create micropost!"
+    destroy:
+      success_mess_deleted: "Micropost deleted."
+      fail_mess_deleted: "Fail to delete micropost!"
+    micropost:
+      deleted_link: "Delete"
+      time_desc_a: "Posted"
+      time_desc_b: "ago."
+      deleted_confirm: "You sure?"
+
```
Comment (manhnt-0960) [inline config/locales/en.yml]: please add newline at end of file

**Comment by EvonneNgXinYi** [comment]
ready

File: app/models/micropost.rb (hunk 3 above)
Comment (EvonneNgXinYi) [reply app/models/micropost.rb:8]: Here I have one extra ques wanted to ask:
image
For this problem, even i uploaded the image in jpeg or png format (VALID image format), but it still showed invalid. May i know how to solve it?

File: app/controllers/microposts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  before_action :correct_user, only: :destroy
+
+  def create
+    @micropost = current_user.microposts.build micropost_params
+    @micropost.image.attach micropost_params[:image]
+    if @micropost.save
+      flash[:success] = t ".success_mess_created"
+    redirect_to root_path
+    else
+      @feed_items = current_user.feed.page(params[:page]).per(Settings.collections.microposts_default_page)
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb]:
```suggestion
      @feed_items = current_user.feed.page(params[:page]).per Settings.collections.microposts_default_page
```

File: app/controllers/static_pages_controller.rb [hunk 6]
```diff
@@ -1,7 +1,14 @@
 class StaticPagesController < ApplicationController
-  def home; end
-  
+  def home
+    return unless logged_in?
+      
+    @micropost = current_user.microposts.build
+    @feed_items = current_user.feed.page(params[:page]).per(Settings.collections.microposts_default_page)
```
Comment (chungpv-1008) [inline app/controllers/static_pages_controller.rb]:
```suggestion
    @feed_items = current_user.feed.page(params[:page]).per Settings.collections.microposts_default_page
```

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -8,7 +8,12 @@ def index
     @users = User.sort_by_name.page(params[:page]).per(Settings.collections.users_default_page)
   end
 
-  def show; end
+  def show
+    @user = User.find_by id: params[:id]
+    return if @user
+    
+    @microposts = @user.microposts.page(params[:page])
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @microposts = @user.microposts.page params[:page]
```

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -70,6 +75,7 @@ def logged_in_user
   end
 
   def correct_user
+    @user = User.find params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
```

File: app/views/microposts/_micropost.html.erb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<li id="micropost-<%= micropost.id %>">
+  <%= link_to gravatar_for(micropost.user, size: Settings.collections.gravatar.size.list), micropost.user %>
+  <span class="user"><%= link_to micropost.user.name, micropost.user %></span>
+  <span class="content">
+    <%= micropost.content %>
+  	<%= image_tag micropost.image if micropost.image.attached? %>
+  </span>
+  <span class="timestamp">
+    <%= t ".time_desc_a" %> <%= time_ago_in_words(micropost.created_at) %> <%= t ".time_desc_b" %>
```
Comment (chungpv-1008) [inline app/views/microposts/_micropost.html.erb]:
```suggestion
    <%= t ".time_desc_a" %> <%= time_ago_in_words micropost.created_at %> <%= t ".time_desc_b" %>
```

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -8,4 +8,13 @@
       </h1>
     </section>
   </aside>
+  <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3><%= t ".prefix_count" %> (<%= @user.microposts.count %>)</h3>
```
Comment (chungpv-1008) [inline app/views/users/show.html.erb]:
```suggestion
      <h3><%= t ".prefix_count" %> (<%= @user.microposts.size %>)</h3>
```

**Comment by EvonneNgXinYi** [comment]
Here I have one extra ques wanted to ask:
image
For this problem, even i uploaded the image in jpeg or png format (VALID image format), but it still showed invalid.But for posting word without image is fine. May i know how to solve it?

**Comment by EvonneNgXinYi** [comment]
ready

File: config/settings.yml [hunk 11]
```diff
@@ -28,4 +30,11 @@ collections:
       default: 80
       list: 50
   users_default_page: 10
+  microposts_default_page: 8
   psword_reset_expired_time_in_hours: 2
+
+image:
+  type: /%w(image/jpeg image/gif image/png)/
```
Comment (saobv-2223) [inline config/settings.yml]:
```suggestion
  type: "image/jpeg image/gif image/png"
```

File: app/models/micropost.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+
+  validates :user_id, presence: true
+  validates :content, presence: true, length: { maximum: Settings.validations.content.max_length }
+  validates :image, content_type: { in: Settings.image.type, message: I18n.t("errors.messages.invalid_image_format") }, size: { less_than: Settings.image.size.megabytes, message: I18n.t("errors.messages.invalid_image_size") }
```
Comment (saobv-2223) [inline app/models/micropost.rb]:
```suggestion
  validates :image, content_type: { in: Settings.image.type.split, message: I18n.t("errors.messages.invalid_image_format") }, size: { less_than: Settings.image.size.megabytes, message: I18n.t("errors.messages.invalid_image_size") }
```

File: app/models/micropost.rb (hunk 3 above)
Comment (saobv-2223) [reply app/models/micropost.rb:8]: you can change such as I suggested

File: app/models/micropost.rb (hunk 12 above)
Comment (EvonneNgXinYi) [reply app/models/micropost.rb]: yes, it work. Thank you

**Comment by EvonneNgXinYi** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**